java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -51,191 +51,211 @@
-    if (pp.i < pp.block.elements().size()) {
-      Tree tree = pp.block.elements().get(pp.i);
-      return tree.is(Tree.Kind.METHOD_INVOCATION);
+    private Stream<JavaFileScannerContext.Location> learnedConstraintFlow(LearnedConstraint learnedConstraint, ExplodedGraph.Edge edge) {
+      Constraint constraint = learnedConstraint.constraint();
+      if (constraint == null || !addToFlow.test(constraint)) {
+        return Stream.empty();
+      }
+      ExplodedGraph.Node parent = edge.parent;
+      Tree nodeTree = parent.programPoint.syntaxTree();
+      if (isMethodInvocationNode(parent)) {
+        return methodInvocationFlow(constraint, edge);
+      }
+      if (nodeTree.is(Tree.Kind.NEW_CLASS)) {
+        return Stream.of(location(parent, String.format("Constructor implies '%s'.", constraint.valueAsString())));
+      }
+      String name = SyntaxTreeNameFinder.getName(nodeTree);
+      String message = name == null ? constraint.valueAsString() : String.format("Implies '%s' is %s.", name, constraint.valueAsString());
+      return Stream.of(location(parent, message));
     }
-    return false;
-  }
 
-  private static int correspondingArgumentIndex(SymbolicValue candidate, ExplodedGraph.Node invocationNode) {
-    MethodInvocationTree mit = (MethodInvocationTree) invocationNode.programPoint.syntaxTree();
-    List<SymbolicValue> arguments = argumentsUsedForMethodInvocation(invocationNode, mit);
-    return arguments.indexOf(candidate);
-  }
+    private Stream<JavaFileScannerContext.Location> methodInvocationFlow(Constraint learnedConstraint, ExplodedGraph.Edge edge) {
+      ExplodedGraph.Node parent = edge.parent;
+      MethodInvocationTree mit = (MethodInvocationTree) parent.programPoint.syntaxTree();
+      Stream.Builder<JavaFileScannerContext.Location> flowBuilder = Stream.builder();
+      SymbolicValue returnSV = edge.child.programState.peekValue();
+      if (symbolicValues.contains(returnSV)) {
+        flowBuilder.add(location(parent, String.format("'%s()' returns %s.", mit.symbol().name(), learnedConstraint.valueAsString())));
+      }
+      SymbolicValue invocationTarget = parent.programState.peekValue(mit.arguments().size());
+      if (symbolicValues.contains(invocationTarget)) {
+        flowBuilder.add(location(parent, "..."));
+      }
+      int argIdx = correspondingArgumentIndex(symbolicValues, parent);
+      if (argIdx != -1) {
+        ExpressionTree argTree = mit.arguments().get(argIdx);
+        String message = String.format("Implies '%s' is %s.", SyntaxTreeNameFinder.getName(argTree), learnedConstraint.valueAsString());
+        flowBuilder.add(new JavaFileScannerContext.Location(message, argTree));
+      }
+      MethodYield selectedMethodYield = edge.child.selectedMethodYield(parent);
+      if (selectedMethodYield != null) {
+        Set<List<JavaFileScannerContext.Location>> xProcFlows = selectedMethodYield.flow(argIdx, domains);
+        // FIXME SONARJAVA-2076 consider all flows from yields
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)